docs/site/Dynamic-models-repositories-controllers.md

```diff
@@ -0,0 +1,218 @@
+        dataLength: 512,
+        dataPrecision: null,
+        dataScale: null,
+        nullable: 'Y',
+      },
+    }
+  },
+  settings: {
+    idInjection: false,
+    mysql: {schema: 'test', table: 'BookStore'},
+  },
+});
+```
+
+### Defining a Model
+
+A LoopBack model class is created by passing a `ModelDefinition` object to
+`@loopback/repository`'s helper function `defineModelClass()`. It expects a
+base model to extend (typically `Model` or `Entity`), folowed by the model
+definition object. In this case it will be `Entity`.
+
+```ts
+const BookModel = defineModelClass(Entity, bookDef);
+```
+
+In case you need to use an existing Model as the base class, you should use the
+`DynamicModelCtor` class, instead of the `defineModelClass()` function.
+
+```ts
+import DynamicModelCtor from '@loopback/repository';
+// Assuming User is a pre-existing Model class in the app
+const StudentModel = new DynamicModelCtor(User, studentDef);
+```
+
+For details about `ModelDefinition`, `defineModelClass`, `Model`, `Entity`, and
+`DynamicModelCtor` refer to the
+[@loopback/repository API documentation](https://loopback.io/doc/en/lb4/apidocs.repository.html).
+
+### Defining a Datasource
+
+Before a repository for this model can be set up, a datasource should be ready.
+If required, set one up dynamically by creating an instance of
+`juggler.DataSource`. It requires a name for the datasource, the connector, and
+a connection url as shown below.
+
+```ts
+const dsName = 'bookstore-ds';
+const bookDs = new juggler.DataSource({
+  name: dsName,
+  connector: require('loopback-connector-mongodb'),
+  url: 'mongodb://sysop:moon@localhost',
+});
+await bookDs.connect();
+app.dataSource(bookDs, dsName);
+```
+
+Connect the datsource by calling its `.connect()` method and attach it to the
+app using `app.dataSource()` method. Note, this method will be available only
+on `RepositoryMixin` apps.
+
+### Defining a Repository
+
+Now that a datasource is ready, a repository can be created for `BookModel`.
+
+The `@loopback/rest-crud` package exports a helper function,
+`defineCrudRepositoryClass`, this function creates a
+`defineCrudRepositoryClass`-based entity CRUD repository class for the given
+model.
+
+```ts
+const BookRepository = defineCrudRepositoryClass(BookModel);
+```
+
+If you want to customize the repository, you can create a copy of
+`defineCrudRepositoryClass`'s
+[implementation](https://github.com/strongloop/loopback-next/blob/00917f5a06ea8a51e1f452f228a6b0b7314809be/packages/rest-crud/src/repository-builder.ts#L25-L47)
+and modify it according to your requirements.
+
+If you want to use a non-`defineCrudRepositoryClass` repository class as the
+base for your repository, you can use the `defineRepositoryClass()` function
+instead.
+
+```ts
+import defineRepositoryClass from '@loopback/repository';
+// Assuming UserRepository is a pre-existing Repository class in the app
+const StudentRepository = defineRepositoryClass(BookModel, UserRepository);
```

Comment:
FYI:

https://github.com/strongloop/loopback-next/blob/20d01a28ed4f192042ff43e2a8dcb3ea98892ab3/packages/repository/src/define-repository-class.ts#L73-L75

Fortunately, if `UserRepository` is not a generic class (which I assume is your intention here), then the compiler infers all types correctly and the comment above does not apply.

```ts
class UserRepository extends DefaultCrudRepository<User, typeof User.prototype.id, UserRelations> {
}

const StudentRepository = defineRepositoryClass(StudentModel, UserRepository);
// StudentRepository has type ModelRepositoryClass<StudentModel, UserRepository>
```

Please add a short note (or a tip?) to provide this explanation to readers too. For example:
1. Explain that in your code snippet, `UserRepository` is not a generic class and therefore the compiler is able to infer everything automatically.
2. If the base class was a generic one, then the user would need to provide both generic parameters (`typeof BookModel` and e.g. `DefaultCrudRepository<BookModel, number, BookRelations>`).